internal/bot/bot.go

```diff
@@ -367,6 +403,12 @@ func (b *Bot) updateLatestCoins(ctx context.Context) error {
 		return err
 	}
 
+	for symbol, coin := range coins {
+		if quoteVolume, ok := b.coinVolumes[symbol]; ok {
+			coin.QuoteVolumeTraded = quoteVolume
+		}
+	}
```

Comment:
This loop runs every time the `updateLatestCoins` function runs, which is frequent. But the volume values stored in this map only change every hour, meaning this loop is unnecessary most of the time. 

Another approach could be to only keep track of the last time the volumes were updated, for instance, and then update them directly in the  `updateVolumeTraded` function:
```go
// updateLatestCoins fetches the latest coins from the market and appends them to the volatilityWindow.
func (b *Bot) updateLatestCoins(ctx context.Context) error {
	b.botLog.Debug("Fetching latest coins.")

	coins, err := b.market.GetCoins(ctx)
	if err != nil {
		return err
	}
	
	// example code, adjust struct and config accordingly
	if b.config.TradingOptions.MinQuoteVolumeTraded != 0.0 &&  b.volumesUpdatedAt.Add(b.config.TradingOptions.VolumeUpdateInterval).Before(time.Now()) {
		b.botLog.Debug("Fetching volume traded of all coins.")

		coinVolumes, err := b.market.GetCoinsVolume(ctx)
		if err != nil {
			return err
		}

		for symbol, coin := range coins {
			if quoteVolume, ok := coinVolumes[symbol]; ok {
				coin.QuoteVolumeTraded = quoteVolume
			}
		}

	}
	
	b.volatilityWindow.AddRecord(coins)

	return nil
}
```

Sorry I didn't realize this approach sooner :sweat_smile:

Reply:
Would this work though? In the `buy` goroutine each time we want to fetch from the API we call `bot.updateLatestCoins`, and inside this function `market.GetCoins` returns coins without the `QuoteVolumeTraded` field, then they get appended to `VolatilityWindow.records`. Immediately after this, in the `buy` function we retrieve the latest record with `VolatilityWindow.IdentifyVolatileCoins`, so most of the time we would retrieve `Coin` structs without a `QuoteVolumeTraded`. We always need `Coin.QuoteVolumeTraded` to be non-null because in the buy logic we call `volatileCoin.Coin.IsAvailableForTrading`.

It's just an O(n) loop with something like 1300 pairs for Binance, I think the performance gain from refactoring this would be negligible.

Reply:
> Would this work though?

Why not exactly? 

> We always need Coin.QuoteVolumeTraded to be non-null because in the buy logic we call volatileCoin.Coin.IsAvailableForTrading.

`updateLatestCoins` is called at the very start of the `buy` function so if the logic were moved there it should still work. 

> It's just an O(n) loop with something like 1300 pairs for Binance, I think the performance gain from refactoring this would be negligible.

True, but if the control flow can be simplified and the bot still functions the exact same way I prefer to pursue it.

Reply:
If you look at the `buy` function, you are also calling `updateLatestCoins` every time you need to refresh the data and fetch from the api, inside the endless `for` loop, not just at the start of the function. With your suggested change, every time except once an hour it will append a `CoinMap` without `Coin.QuoteVolumeTraded` data to the `VolatilityWindow.records`, and that will break the `buy` function check `volatileCoin.Coin.IsAvailableForTrading` in my PR, so at every data refresh we might enter positions that are supposed to be filtered out.

Reply:
Ehhh I'll just approve the PR as is, I'm not really in the right headspace right now to break my head over it. If changes are needed in the future I'll do so in another PR (and assign you for review).